Only don't do custom rendering if it would cause a crash

pupnewfster opened the pull request:
Makes it so https://github.com/aidancbrady/Mekanism/issues/5293 doesn't happen but unlike the implementation in https://github.com/aidancbrady/Mekanism/commit/fe9a0e660077cc8f374aabfca029f30a1582495e it does not cause the items to turn invisible with `allowEmissiveItems=true`. The items do render as invisible in the in world viewing of the tinkers construct crafting table, but you can see the items properly when looking in the GUI.

The linked commit makes it not do the custom rendering for items at all if "the Tesselator isDrawing." (copied quote from the commit message) so any items that need custom rendering don't get it done.

pupnewfster committed:
Only don't do custom rendering if it would cause a crash

pupnewfster committed:
Only register JEI categories if the machine that they are for is enabled.
Fixes https://github.com/aidancbrady/Mekanism/issues/5197

dizzyd commented on src/main/java/mekanism/client/render/item/BakedCustomItemModel.java:
```diff
@@ -375,38 +377,44 @@ private void doRender(TransformType type) {
 
             return faceQuads;
         }
+        boolean doCustomRendering = true;
```
Can you please provide some commentary here on why GLContext.getCapabilities will crash and how we're using that test to determine how to handle the render?

Also, a little concerned about doing this try/catch every render cycle...

pupnewfster replied:
`GlStateManager.translate` calls `GL11.glTranslatef`
The code for this method starts as
`ContextCapabilities caps = GLContext.getCapabilities();`
If it fails to be able to get the capabilities, then it will throw a runtime exception there.

I was not able to figure out why GLContext sometimes has null capabilities, but from what I can tell it has something to do with CodeChickenLib having its own BlockRendererDispatcher. I BELIEVE this makes it so that when the custom rendering happens for mekanism (which renders some items as blocks), gets called by TConstruct to render it in the world, it flips out potentially because of already rendering the block itself and not knowing how to render it internally. Still not sure why it would have null capabilities though.

The reason I pulled this out is that then it is easier to keep track of if it was running before when we paused it. I can do a little testing if you would rather, and try to make the try/catch be put around the pause/resume renderer. That way it doesn't make an extra check to `getCapabilities()`, though it may be slightly more messy. Not sure if I would have to pop the matrix regardless when it crashes or not. The other option which I am much more open to is to only do that try/catch if allowEmissiveItems is set to true by forge. (As this is the only case where I have seen it crash)

dizzyd replied:
Yes - let's do the allowEmissiveitems check around the try/catch and see if that yields a similar result; the performance _should_ be better and it also makes the intent of the code more clear.

It would also be great if the commentary above could be in a comment in the code for future posterity (and also because I have the memory of a bird when it comes to complex code bits) :)

pupnewfster replied:
One other thing to note, which I forgot to mention while writing the description... That if we do add that allowEmissiveItems check, we have to bump the min req forge version to 2781

Edit: also it appears it crashes on the pushMatrix, because that also ends up calling getCapabilites(), it doesn't get to the `translate`

pupnewfster committed:
Fixes crash with tinkers/CCL/allowEmissiveItems in a better way.
This also makes it so that it renders the failed items as null textured in the tinkers table instead of them being invisible.

pupnewfster committed:
Kepe track of which renderstate got paused

pupnewfster committed:
Merge

pupnewfster commented:
Did a much better fix, which should have  better compat with other mods and stuff. Basically if rendering the item will cause a crash, it says that it does in fact use built in rendering so that the custom rendering code does not break it.

The commit about JEI can be ignored, I by mistake had committed that to my local branch, and the merge commit makes it so that is ignored as that is already in master.

Shadows-of-Fire commented:
Just looking at that stuff... That entire model class is a giant hack, and should *really* be a TEISR.  In fact, that'd probably resolve all your issues anyway.

You should never actually do anything in a getter method.  This implementation should be deleted as soon as physically possible.

pupnewfster commented:
Oh we know it is a massive hack, scrapping the old model system is pretty high up on our list of things to do. Given the entire model system is a hack.
